spec/unit/invoca/metrics/prometheus/metrics_registry_spec.rb

```
@@ -0,0 +1,70 @@
+# frozen_string_literal: true
+
+require_relative '../../../../spec_helper'
+
+describe Invoca::Metrics::Prometheus::MetricsRegistry do
+  let(:metrics_registry) { described_class.new }
+  let(:metric_name) { :dynamic_method }
+  let(:metric) { :metric }
+  let(:metric_2) { :metric_2 }
+
+  describe '#register' do
+    before { metrics_registry.register(metric_name, metric) }
+    context 'valid name is provided' do
+      it 'adds a metric to the registry' do
+        expect(metrics_registry.metrics[metric_name]).to eq(metric)
+      end
+
+      it 'can access method with dot operator' do
+        expect(metrics_registry.dynamic_method).to eq(metric)
+      end
+    end
+
+    context 'duplicate name is provided' do
+      subject { metrics_registry.register(metric_name, metric_2) }
+
+      # TODO: Do we actually want this to raise instead of warn?
```

Should we `warn` and overwrite the duplicate metric name or fail fast? After putting this together I'm leaning towards failing fast to ensure no metrics are overwritten

---

I'm ok with either, but I'm inclined to agree that fail fast is probably better here. I don't think there are legit cases where we'd overwrite one with another.

But either way, didn't we agree to only warn/fail if the registration is different? I think it will be ok to have redundant registrations from different source classes. An example we talked about in a recent meaning would be Ringswitch classes that generate incalls vs. outcalls. They might both register a `call` metric with a `direction:` label. In graphite-based systems those look like different metrics, but Prometheus encourages the same metric (base) name with different labels.

Which points out: you probably should remember the source that registered it, so that you can point to it in the exception/warning message. The class that registered it is probably sufficient. I suppose you could remember the entire `caller` but that could be a bit much to show in the logs!